packages/flutter/lib/src/material/app.dart

```diff
@@ -1173,3 +1166,115 @@ class _MaterialAppState extends State<MaterialApp> {
     );
   }
 }
+
+enum _MaterialInspectorButtonVariant { filled, toggle, iconOnly }
+
+class _MaterialInspectorButton extends StatelessWidget {
+  const _MaterialInspectorButton({
+    required this.onPressed,
+    required this.semanticLabel,
+    required this.icon,
+    required this.isDarkTheme,
+    this.buttonKey,
+  }) : _variant = _MaterialInspectorButtonVariant.filled,
+       _toggledOn = null;
+
+  const _MaterialInspectorButton.toggle({
+    required this.onPressed,
+    required this.semanticLabel,
+    required this.icon,
+    required this.isDarkTheme,
+    bool toggledOn = true,
+  }) : buttonKey = null,
+       _variant = _MaterialInspectorButtonVariant.toggle,
+       _toggledOn = toggledOn;
+
+  const _MaterialInspectorButton.iconOnly({
+    required this.onPressed,
+    required this.semanticLabel,
+    required this.icon,
+    required this.isDarkTheme,
+  }) : buttonKey = null,
+       _variant = _MaterialInspectorButtonVariant.iconOnly,
+       _toggledOn = null;
+
+  final VoidCallback onPressed;
+  final String semanticLabel;
+  final IconData icon;
+  final bool isDarkTheme;
+  final GlobalKey? buttonKey;
+  final _MaterialInspectorButtonVariant _variant;
+  final bool? _toggledOn;
+
+  static const double _buttonSize = 32.0;
+  static const double _buttonIconSize = 18.0;
+  static const EdgeInsets _buttonPadding = EdgeInsets.zero;
+  static const BoxConstraints _buttonConstraints = BoxConstraints.tightFor(
+    width: _buttonSize,
+    height: _buttonSize,
+  );
+
+  @override
+  Widget build(BuildContext context) {
+    return IconButton(
+      key: buttonKey,
+      onPressed: onPressed,
+      iconSize:
+          _variant == _MaterialInspectorButtonVariant.iconOnly ? _buttonSize : _buttonIconSize,
```

Comment:
general question: is there a way to share some of this code between the material and cupertino implementations? we seem to be duplicating several things in both places. For example, we are duplicating the variant enums and some of the button logic

Reply:
Refactored to share some of the code between the two